govcd/vapp.go

```diff
@@ -786,6 +798,328 @@ func (vapp *VApp) AddIsolatedNetwork(newIsolatedNetworkSettings *VappNetworkSett
+}
+
+// UpdateNetworkAsync asyncronously updates vApp networks (isolated or connected to Org VDC network).
+// Returns task or error
+func (vapp *VApp) UpdateNetworkAsync(networkSettingsToUpdate *VappNetworkSettings, orgNetwork *types.OrgVDCNetwork) (Task, error) {
+	util.Logger.Printf("[TRACE] UpdateNetworkAsync with values: %#v and connect to org network: %#v", networkSettingsToUpdate, orgNetwork)
+	currentNetworkConfiguration, err := vapp.GetNetworkConfig()
+	if err != nil {
+		return Task{}, err
+	}
+	var networkToUpdate types.VAppNetworkConfiguration
+	var networkToUpdateIndex int
+	for index, networkConfig := range currentNetworkConfiguration.NetworkConfig {
+		uuid, err := GetUuidFromHref(networkConfig.Link.HREF)
+		if err != nil {
+			return Task{}, err
+		}
+
+		if uuid == networkSettingsToUpdate.Id {
+			networkToUpdate = networkConfig
+			networkToUpdateIndex = index
+			break
+		}
+	}
+
+	if networkToUpdate == (types.VAppNetworkConfiguration{}) {
+		return Task{}, fmt.Errorf("not found network to update with Id %s", networkSettingsToUpdate.Id)
+	}
+
+	if networkToUpdate.Configuration.Features.NatService != nil && networkSettingsToUpdate.NatEnabled != nil {
+		networkToUpdate.Configuration.Features.NatService.IsEnabled = *networkSettingsToUpdate.NatEnabled
+	}
+	if networkToUpdate.Configuration.Features.NatService == nil && networkSettingsToUpdate.NatEnabled != nil {
+		networkToUpdate.Configuration.Features.NatService = &types.NatService{IsEnabled: *networkSettingsToUpdate.NatEnabled, NatType: "ipTranslation", Policy: "allowTrafficIn"}
+	}
+
+	if networkToUpdate.Configuration.Features.FirewallService != nil && networkSettingsToUpdate.FirewallEnabled != nil {
+		networkToUpdate.Configuration.Features.FirewallService.IsEnabled = *networkSettingsToUpdate.FirewallEnabled
+	}
+	if networkToUpdate.Configuration.Features.FirewallService == nil && networkSettingsToUpdate.FirewallEnabled != nil {
+		networkToUpdate.Configuration.Features.FirewallService = &types.FirewallService{IsEnabled: *networkSettingsToUpdate.FirewallEnabled}
+	}
+
+	networkToUpdate.Configuration.Features.FirewallService.IsEnabled = *networkSettingsToUpdate.FirewallEnabled
+	networkToUpdate.Configuration.RetainNetInfoAcrossDeployments = networkSettingsToUpdate.RetainIpMacEnabled
+	if networkToUpdate.Configuration.ParentNetwork != nil && orgNetwork != nil && networkToUpdate.Configuration.ParentNetwork.HREF != orgNetwork.HREF {
+		networkToUpdate.Configuration.ParentNetwork = &types.Reference{HREF: orgNetwork.HREF}
+	}
+	if networkToUpdate.Configuration.ParentNetwork != nil && orgNetwork == nil {
+		networkToUpdate.Configuration.ParentNetwork = nil
+	}
+	networkToUpdate.Description = networkSettingsToUpdate.Description
+	networkToUpdate.NetworkName = networkSettingsToUpdate.Name
+	networkToUpdate.Configuration.GuestVlanAllowed = networkSettingsToUpdate.GuestVLANAllowed
+	networkToUpdate.Configuration.IPScopes.IPScope[0].Gateway = networkSettingsToUpdate.Gateway
```

Comment:
Are we not binding ourselves to first object while many IP ranges can exist?

Reply:
Well, in UI you can't have more gateways, etc. So if you know how to verify need of that when we consider changing a lot this PR.